hbase-it/src/test/java/org/apache/hadoop/hbase/ShellExecCoproc.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.hadoop.hbase.coprocessor.protobuf.generated.ShellExecEndpoint.ShellExecRequest;
+import org.apache.hadoop.hbase.coprocessor.protobuf.generated.ShellExecEndpoint.ShellExecResponse;
+import org.apache.hadoop.hbase.ipc.CoprocessorRpcUtils;
+import org.apache.hadoop.util.Shell;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.hbase.thirdparty.com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Receives shell commands from the client and executes them blindly.
+ */
+@InterfaceAudience.Private
+public class ShellExecCoproc
+  extends ShellExecEndpoint.ShellExecService
+  implements MasterCoprocessor, RegionServerCoprocessor {
+  private static final Logger LOG = LoggerFactory.getLogger(ShellExecCoproc.class);
+
+  public static final String ASYNC_DELAY_KEY = "hbase.it.shellexeccoproc.async.delay";
+  public static final long DEFAULT_ASYNC_DELAY = 1_000;
+
+  private final ExecutorService workerPool;
+  private Configuration conf;
+
+  public ShellExecCoproc() {
+    workerPool = Executors.newSingleThreadExecutor(
+      new ThreadFactoryBuilder()
+        .setNameFormat(ShellExecCoproc.class.getSimpleName() + "-{}")
+        .setDaemon(true)
+        .setUncaughtExceptionHandler((t, e) -> LOG.warn("Thread {} threw", t, e))
+        .build());
+  }
+
+  @Override
+  public Iterable<Service> getServices() {
+    return Collections.singletonList(this);
+  }
+
+  @Override
+  public void start(CoprocessorEnvironment env) {
+    conf = env.getConfiguration();
+  }
+
+  @Override
+  public void shellExec(
+    final RpcController controller,
+    final ShellExecRequest request,
+    final RpcCallback<ShellExecResponse> done
+  ) {
+    final Shell.ShellCommandExecutor shell = prepareShell(request, false);
+
+    ShellExecResponse.Builder builder = ShellExecResponse.newBuilder();
+    try {
+      builder = doExec(shell, builder);
+    } catch (IOException e) {
+      LOG.info("Failure launching process", e);
+      CoprocessorRpcUtils.setControllerException(controller, e);
+    }
+
+    done.run(builder.build());
+  }
+
+  @Override
+  public void shellExecAsync(
+    final RpcController controller,
+    final ShellExecRequest request,
+    final RpcCallback<ShellExecResponse> done
+  ) {
+    final Shell.ShellCommandExecutor shell = prepareShell(request, true);
+    final long sleepDuration = conf.getLong(ASYNC_DELAY_KEY, DEFAULT_ASYNC_DELAY);
+    workerPool.submit(() -> {
+      try {
+        // sleep first so that the RPC can ACK.
+        Thread.sleep(sleepDuration);
+        doExec(shell, ShellExecResponse.newBuilder());
+      } catch (InterruptedException e) {
+        LOG.info("Interrupted before launching process.", e);
+      } catch (IOException e) {
+        LOG.info("Failure launching process", e);
+      }
+    });
+    done.run(ShellExecResponse.newBuilder().build());
```

Comment:
This doesn't look right to me, shouldn't this be predicated on whether the thread above is successful? (You probably need to chain them?.. thenApply()...)

Reply:
See my earlier comment. This isn't async in the `CompletableFuture` sense. It's a call without a response.